lib/ClangImporter/ImportType.cpp

```diff
@@ -590,12 +602,25 @@ namespace {
         // suppressed. Treat it as a typedef.
         return None;
       }
+
       if (index > genericSig->getGenericParams().size()) {
         return ImportResult();
       }
+
       auto *paramDecl = genericSig->getGenericParams()[index];
-      return ImportResult(genericEnv->mapTypeIntoContext(paramDecl),
-                          ImportHint::ObjCPointer);
+      auto type = genericEnv->mapTypeIntoContext(paramDecl);
+
+      // Check if parameter's declaration context is the same as
+      // current declaration context, otherwise we can't use type
+      // type without re-substituting it. This might happen when
+      // Objective-C `typedef` is declarated inside of interface
+      // and tries to use its generic parameters.
+      if (auto *ownerDC = genericEnv->getOwningDeclContext()) {
+        if (DC && DC != ownerDC && !ownerDC->isChildContextOf(DC))
+          type = mapTypeIntoContext(ownerDC, DC, type);
```

Comment:
This seems more complicate than necessary... can't we just `DC->mapTypeIntoContext(type)`  once, and not bother doing the `DC != ownerDC` check?